docs/rules/no-mutation-props.md

```diff
@@ -0,0 +1,19 @@
+# Prevent mutation of this.props (no-mutation-props)
+
+NEVER mutate `this.props`, as all React components must act like pure functions with respect to their props. 
+Treat `this.props` as if it were immutable. More info available at [https://facebook.github.io/react/docs/components-and-props.html#props-are-read-only](https://facebook.github.io/react/docs/components-and-props.html#props-are-read-only)
+
+## Rule Details
+
+This rule is aimed to forbid the use of mutating `this.props`.
+
+The following patterns are considered warnings:
+
+```jsx
+var Hello = React.createClass({
+  render: function() {
+    this.props.name = this.props.name.toUpperCase();
```

Comment:
What about things like `this.props.list.push('foo')`, or other array mutators? What about `delete this.props.name`, or `Object.defineProperty(this.props, something)`, or `Object.assign(this.props, something)`?

Reply:
I've got a PR submitted that builds on @ianschmitz's work and catches this and many other things as well :) https://github.com/yannickcr/eslint-plugin-react/pull/1416